pgvectorscale/src/access_method/distance_aarch64.rs

```diff
@@ -0,0 +1,222 @@
+//! Calculate the distance by vector arithmetic optimized for aarch64 neon intrinsics
+
+use core::arch::aarch64::{self, *};
+
+#[cfg(not(target_feature = "neon"))]
+#[cfg(not(doc))]
+compile_error!(
+    "On arm, the neon feature must be enabled. Set RUSTFLAGS=\"-C target-feature=+neon\""
+);
+
+// Eventually would like to upstream changed into simdeez, so to simplify the transition,
+// creating this struct for an easy refactor
+struct S;
+
+impl S {
+    const VF32_WIDTH: usize = 4; // 128bit register for ARM NEON
+
+    fn setzero_ps() -> float32x4_t {
+        let x: float32x4_t;
+        unsafe {
+            let zero: f32 = 0.0;
+            x = aarch64::vld1q_dup_f32(&zero);
+        }
+        x
+    }
+
+    fn loadu_ps(a: &f32) -> float32x4_t {
+        unsafe { aarch64::vld1q_f32(a) }
+    }
+
+    fn horizontal_add_ps(a: float32x4_t) -> f32 {
+        unsafe { aarch64::vaddvq_f32(a) }
+    }
+
+    fn fmadd_ps(a: float32x4_t, b: float32x4_t, c: float32x4_t) -> float32x4_t {
+        unsafe { aarch64::vfmaq_f32(c, a, b) }
+    }
+
+    fn mult_ps(a: float32x4_t, b: float32x4_t) -> float32x4_t {
+        unsafe { aarch64::vmulq_f32(a, b) }
+    }
+
+    fn add_ps(a: float32x4_t, b: float32x4_t) -> float32x4_t {
+        unsafe { aarch64::vaddq_f32(a, b) }
+    }
+
+    fn sub_ps(a: float32x4_t, b: float32x4_t) -> float32x4_t {
+        unsafe { aarch64::vsubq_f32(a, b) }
+    }
+}
+
+pub fn distance_l2_aarch64_neon(x: &[f32], y: &[f32]) -> f32 {
+    let mut accum0 = S::setzero_ps();
```

Comment:
Seeking feedback: the body of this function is the same as the _x86 version right? (Just with a different S?).

Can we somehow refactor so that both use a common macro/macro_rule?

Reply:
Yeah great idea, I'll investigate. Also, the only thing holding back the distance_cosine function is defining the +,-,* operators for `S`, so there's potential to do both l2 and cosine via shared macros